pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java

```diff
@@ -1083,4 +1117,103 @@ private void createPendingLoadTopic() {
         }
 
     }
+    
+    /**
+     * Adds given dispatcher's unackMessage count to broker-unack message count and if it reaches to the
+     * {@link #maxUnackedMessages} then it blocks all the dispatchers which has unack-messages higher than
+     * {@link #maxUnackedMsgsPerDispatcher}. It unblocks all dispatchers once broker-unack message counts decreased to
+     * ({@link #maxUnackedMessages}/2)
+     * 
+     * @param dispatcher
+     * @param numberOfMessages
+     */
+    public void addUnAckedMessages(PersistentDispatcherMultipleConsumers dispatcher, int numberOfMessages) {
+        // don't block dispatchers if maxUnackedMessages = 0
+        if (maxUnackedMessages <= 0) {
+            return;
+        }
+        int unAckedMessages = TOTAL_UNACKED_MESSAGES_UPDATER.addAndGet(this, numberOfMessages);
```

Comment:
This counter might become a contention point since it's shared across all the instances and threads.

What about using a `LongAdder` instead (it has per-thread caches of the counters)? Then, periodically, we could check the `sum()` of the adder and take action if needed.